cooked-validators/src/Cooked/Tx/Constraints.hs

```diff
@@ -27,78 +29,132 @@ import qualified PlutusTx as Pl
 
--- | Map from datum hashes to string representation of all the datums carried.
--- We use this in order to display data to the use when testing. Its often
--- easier to read the original datatype that was placed into a UTxO
--- instead of its respective @toBuilinData@ image.
-extractDatumStrFromConstraint :: Constraint -> M.Map Pl.DatumHash String
-extractDatumStrFromConstraint (PaysScript _validator datumsAndValues) =
-  M.fromList
-    . map ((\d -> (Pl.datumHash . Pl.Datum $ Pl.toBuiltinData d, show d)) . fst)
-    $ datumsAndValues
-extractDatumStrFromConstraint (SpendsScript _validator _redeemer (_out, datum)) =
-  M.singleton (Pl.datumHash . Pl.Datum $ Pl.toBuiltinData datum) (show datum)
-extractDatumStrFromConstraint (PaysPKWithDatum _pk _stak mdat _v) =
-  maybe M.empty (\d -> M.singleton (Pl.datumHash . Pl.Datum $ Pl.toBuiltinData d) (show d)) mdat
-extractDatumStrFromConstraint _ = M.empty
-
--- | Converts our constraint into a 'LedgerConstraint',
---  which later can be used to generate a transaction. The universally
---  quantified type-variable is there on purpose, to enable us to
---  easily spend from multiple scripts at the same time.
-toLedgerConstraint :: Constraint -> LedgerConstraint a
-toLedgerConstraint (SpendsScript v r ((oref, o), _a)) = (lkups, constr)
-  where
-    lkups =
-      Pl.otherScript (Pl.validatorScript v)
-        <> Pl.unspentOutputs (M.singleton oref o)
-    constr = Pl.mustSpendScriptOutput oref (Pl.Redeemer $ Pl.toBuiltinData r)
-toLedgerConstraint (PaysScript v outs) = (lkups, constr)
-  where
-    lkups = Pl.otherScript (Pl.validatorScript v)
-    constr = mconcat $
-      flip map outs $ \(d, val) ->
-        Pl.mustPayToOtherScript
-          (Pl.validatorHash $ Pl.validatorScript v)
-          (Pl.Datum $ Pl.toBuiltinData d)
-          val
-toLedgerConstraint (PaysPKWithDatum p stak dat v) = (lkups, constr)
-  where
-    mData = fmap (Pl.Datum . Pl.toBuiltinData) dat
-
-    lkups =
-      maybe mempty Pl.otherData mData
-        -- TODO: do we want to akk ownStakePubKeyHash on 'PaysPKWithDatum'? Would we rather have
-        -- a different 'WithOwnStakePubKeyHash' constraint?
-        <> maybe mempty Pl.ownStakePubKeyHash stak
-    constr = Pl.singleton $ Pl.MustPayToPubKeyAddress (Pl.PaymentPubKeyHash p) stak mData v
-toLedgerConstraint (SpendsPK (oref, o)) = (lkups, constr)
-  where
-    lkups = Pl.unspentOutputs (M.singleton oref o)
-    constr = Pl.mustSpendPubKeyOutput oref
-toLedgerConstraint (Mints Nothing pols v) = (lkups, constr)
-  where
-    lkups = foldMap Pl.mintingPolicy pols
-    constr = Pl.mustMintValue v
-toLedgerConstraint (Mints (Just r) pols v) = (lkups, constr)
-  where
-    lkups = foldMap Pl.mintingPolicy pols
-    constr = Pl.mustMintValueWithRedeemer (Pl.Redeemer (Pl.toBuiltinData r)) v
-toLedgerConstraint (Before t) = (mempty, constr)
+-- | Class for common operations on output and misc constraints.
+class Constraint constraint where
```

Comment:
I'm not sure a typeclass makes sense here. Do we plan to create more types that should satisfy `Constraint`? That being said, I'm ok leaving it as a typeclass but we need a different name and comment. Something like `class InternalConstrFunctions` and a comment explaining that a user never has to worry about this class is important.

Reply:
We may have an `InConstraint` datatype if at some point in the future we can order inputs in a transaction as well. Having a typeclass and common functions felt more natural in parts of the code they are used. It hides the split of constraints into `Misc` and `Out` in parts of the code there is no reason to care about that.

I agree on naming it differently. Especially since `Constraints` (plural) also instantiates it.

Reply:
Makes sense; then lets keep it a typeclass and just rename it

Reply:
Renamed and explained